src/ncp/ncp_uart.cpp

```diff
@@ -297,6 +314,74 @@ void NcpUart::HandleError(otError aError, uint8_t *aBuf, uint16_t aBufLength)
     otNcpStreamWrite(0, reinterpret_cast<uint8_t *>(hexbuf + 1), static_cast<int>(strlen(hexbuf) - 1));
 }
 
+#if OPENTHREAD_ENABLE_NCP_SPINEL_TRANSFORMER
+
+NcpUart::NcpFrameBufferTransformerReader::NcpFrameBufferTransformerReader(NcpFrameBuffer &aTxFrameBuffer) :
+    mTxFrameBuffer(aTxFrameBuffer),
+    mInputBufferLength(0),
+    mOutputBufferReadIndex(0),
+    mOutputBufferLength(0) {}
+
+bool NcpUart::NcpFrameBufferTransformerReader::IsEmpty() const
+{
+    return mTxFrameBuffer.IsEmpty() && !mOutputBufferLength;
+}
+
+otError NcpUart::NcpFrameBufferTransformerReader::OutFrameBegin()
+{
+    otError status = OT_ERROR_FAILED;
+
+    Reset();
+
+    if ((status = mTxFrameBuffer.OutFrameBegin()) == OT_ERROR_NONE)
+    {
+        while (!mTxFrameBuffer.OutFrameHasEnded())
```

Comment:
I think the idea here is to read the entire frame from `mTxFrameBuffer` into `mInputBuffer` and then transform it. We can simplify the whole loop with the following (read the entire frame in one go instead of byte by byte). 
```
mInputBufferLength = mTxFrameBuffer.OutFrameGetLength();
assert(mInputBufferLength <= sizeof(mInputBuffer));
mTxFrameBuffer.OutFrameRead(mInputBuffer, mInputBufferLength);
```

Reply:
Done. Thanks for suggestions.